osu.Game.Rulesets.Mania/UI/ManiaTouchInputArea.cs

```diff
@@ -0,0 +1,227 @@
+        public override bool PropagateNonPositionalInputSubTree => true;
+
+        [SettingSource("Spacing", "The spacing between receptors.")]
+        public BindableFloat Spacing { get; } = new BindableFloat(10)
+        {
+            Precision = 1,
+            MinValue = 0,
+            MaxValue = 100,
+        };
+
+        [SettingSource("Opacity", "The receptor opacity.")]
+        public BindableFloat Opacity { get; } = new BindableFloat(1)
+        {
+            Precision = 0.1f,
+            MinValue = 0,
+            MaxValue = 1
+        };
+
+        [Resolved]
+        private ManiaRulesetConfigManager rulesetConfig { get; set; } = null!;
+
+        private GridContainer gridContainer = null!;
+
+        private readonly BindableBool touchControls = new BindableBool();
+
+        public ManiaTouchInputArea(DrawableManiaRuleset drawableRuleset)
+        {
+            this.drawableRuleset = drawableRuleset;
+
+            Anchor = Anchor.BottomCentre;
+            Origin = Anchor.BottomCentre;
+
+            RelativeSizeAxes = Axes.Both;
+            Height = 0.5f;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            List<Drawable> receptorGridContent = new List<Drawable>();
+            List<Dimension> receptorGridDimensions = new List<Dimension>();
+
+            bool first = true;
+
+            foreach (var stage in drawableRuleset.Playfield.Stages)
+            {
+                foreach (var column in stage.Columns)
+                {
+                    if (!first)
+                    {
+                        receptorGridContent.Add(new Gutter { Spacing = { BindTarget = Spacing } });
+                        receptorGridDimensions.Add(new Dimension(GridSizeMode.AutoSize));
+                    }
+
+                    receptorGridContent.Add(new ColumnInputReceptor
+                    {
+                        Action = { BindTarget = column.Action },
+                        Enabled = { BindTarget = touchControls },
+                    });
+                    receptorGridDimensions.Add(new Dimension());
+
+                    first = false;
+                }
+            }
+
+            InternalChild = gridContainer = new GridContainer
+            {
+                RelativeSizeAxes = Axes.Both,
+                AlwaysPresent = true,
+                Content = new[] { receptorGridContent.ToArray() },
+                ColumnDimensions = receptorGridDimensions.ToArray()
+            };
+        }
+
+        private IBindable<ManiaMobilePlayStyle> mobilePlayStyle = null!;
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            mobilePlayStyle = rulesetConfig.GetBindable<ManiaMobilePlayStyle>(ManiaRulesetSetting.MobilePlayStyle);
+            mobilePlayStyle.BindValueChanged(p => touchControls.Value = p.NewValue == ManiaMobilePlayStyle.TouchControls, true);
```

Comment:
It looks like the touch controls will get stuck on the screen if the user disables them halfway through a play.

Also they don't seem to show up on desktop, and I'm not sure why. As far as I'm reading they should until the first key press.

Reply:
They must be hidden by default, it's because this component inherits from `VisibilityContainer` which hides by default.

Reply:
Please re-read. The case I'm touching on is where the user *disables* them during gameplay.